Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -1010,35 +999,103 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+        }
     }
 
     m_canvas.drawRect(destRect, paint);
 }
 
-void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& patternTransform, const FloatPoint& phase, const FloatSize& spacing, ImagePaintingOptions options)
+void GraphicsContextSkia::beginRecording()
 {
-    auto* nativeImageGrContext = nativeImage.backend().skiaGrContext();
-    if (nativeImageGrContext && nativeImageGrContext != PlatformDisplay::sharedDisplay().skiaGrContext()) {
-        // If we encounter an accelerated NativeImage (skiaGrContext() != nullptr), we are in the threaded GPU rendering painting mode, verify that.
-        ASSERT(m_renderingMode == RenderingMode::Accelerated);
-        ASSERT(ProcessCapabilities::canUseAcceleratedBuffers());
-#if USE(COORDINATED_GRAPHICS)
-        ASSERT(SkiaPaintingEngine::numberOfGPUPaintingThreads() > 0);
-#endif
-        // The 'nativeImage' was produced on another thread -- to use it here, we need to create a new NativeImage, that wraps the existing GPU resource.
-        if (auto newNativeImage = nativeImage.backend().copyAcceleratedNativeImageBorrowingBackendTexture())
-            drawSkiaPattern(newNativeImage->platformImage(), newNativeImage->size(), destRect, tileRect, patternTransform, phase, spacing, options);
-        return;
+    ASSERT(m_contextMode == ContextMode::PaintingMode);
+    m_contextMode = ContextMode::RecordingMode;
+}
+
+SkiaImageToFenceMap GraphicsContextSkia::endRecording()
+{
+    ASSERT(m_contextMode == ContextMode::RecordingMode);
+    m_contextMode = ContextMode::PaintingMode;
+    return WTFMove(m_imageToFenceMap);
+}
+
+template<typename T>
+inline std::unique_ptr<GLFence> createAcceleratedRenderingFence(T object)
+{
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
+    if (!glContext || !glContext->makeContextCurrent())
+        return nullptr;
+
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+
+    grContext->flush(object);
+
+    if (GLFence::isSupported()) {
+        grContext->submit(GrSyncCpu::kNo);
+
+        if (auto fence = GLFence::create())
+            return fence;
     }
 
-    drawSkiaPattern(nativeImage.platformImage(), nativeImage.size(), destRect, tileRect, patternTransform, phase, spacing, options);
+    grContext->submit(GrSyncCpu::kYes);
+    return nullptr;
+}
+
+std::unique_ptr<GLFence> GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded(SkSurface* surface)
+{
+    if (!surface || !surface->recordingContext())
+        return nullptr;
+    return createAcceleratedRenderingFence<decltype(surface)>(surface);
+}
+
+std::unique_ptr<GLFence> GraphicsContextSkia::createAcceleratedRenderingFenceIfNeeded(const sk_sp<SkImage>& image)
+{
+    if (!image || !image->isTextureBacked())
+        return nullptr;
+    return createAcceleratedRenderingFence<decltype(image)>(image);
```

Comment:
why is `decltype` used here rather than the actual type, since we know it?

Reply:
Leftover from a previous templatified version, good catch.

Reply:
Note, your comment applies to the PR #42843, this PR just happens to contain the same commit as pre-requisite, will fix your comment in the other PR, and rebase this one, once the other one landed.